File: scripts/differential/java/T16R.java
package app.freerouting.geometry.planar;

public class T16R {
  static StringBuilder sb = new StringBuilder();
  static long seed;

  static long next() { seed ^= seed << 13; seed ^= seed >>> 7; seed ^= seed << 17; return seed; }
  static int r(int lo, int hi) { long v = (next() >>> 1) % (long) (hi - lo + 1); return (int) (lo + v); }

  interface Th { String get() throws Throwable; }
  static void emit(String name, Th t) {
    String v;
    try { v = t.get(); } catch (Throwable e) { v = "EXC"; }
    sb.append(name).append("=").append(v).append("\n");
  }

  static String f(double d) { return "b" + Double.doubleToRawLongBits(d); }
  static String cpt(Point p) {
    if (p == null) return "null";
    if (p instanceof IntPoint q) return "I" + q.x + "/" + q.y;
    RationalPoint rr = (RationalPoint) p;
    return "R" + rr.x + "/" + rr.y + "/" + rr.z;
  }
  static String fp(FloatPoint p) { return p == null ? "null" : "(" + f(p.x) + "," + f(p.y) + ")"; }
  static String ln(Line l) { return l == null ? "null" : "[" + cpt(l.a) + ";" + cpt(l.b) + "]"; }
  static String box(IntBox b) { return b == null ? "null" : "B(" + b.ll.x + "," + b.ll.y + "," + b.ur.x + "," + b.ur.y + ")"; }
  static String oct(IntOctagon o) {
    return o == null ? "null" : "O(" + o.leftX + "," + o.bottomY + "," + o.rightX + "," + o.topY + ","
        + o.upperLeftDiagonalX + "," + o.lowerRightDiagonalX + "," + o.lowerLeftDiagonalX + "," + o.upperRightDiagonalX + ")";
  }
  static String ts(TileShape s) {
    if (s == null) return "null";
    if (s instanceof IntBox b) return box(b);
    if (s instanceof IntOctagon o) return oct(o);
    StringBuilder b = new StringBuilder("S{");
    for (int i = 0; i < s.borderLineCount(); i++) { if (i > 0) b.append(" "); b.append(ln(s.borderLine(i))); }
    return b.append("}").toString();
  }
  static String tsArr(TileShape[] a) {
    if (a == null) return "null";
    StringBuilder b = new StringBuilder("[" + a.length + ":");
    for (TileShape s : a) { b.append(ts(s)).append(";"); }
    return b.append("]").toString();
  }
  static String pl(Polyline p) {
    if (p == null) return "null";
    StringBuilder b = new StringBuilder("P[" + p.lines.length + ":");
    for (Line l : p.lines) b.append(ln(l)).append(";");
    return b.append("]").toString();
  }
  static String plArr(Polyline[] a) {
    if (a == null) return "null";
    StringBuilder b = new StringBuilder("[" + a.length + ":");
    for (Polyline p : a) b.append(pl(p)).append(";");
    return b.append("]").toString();
  }
  static String ptArr(Point[] a) {
    if (a == null) return "null";
    StringBuilder b = new StringBuilder("[" + a.length + ":");
    for (Point p : a) b.append(cpt(p)).append(";");
    return b.append("]").toString();
  }
  static String fpArr(FloatPoint[] a) {
    if (a == null) return "null";
    StringBuilder b = new StringBuilder("[" + a.length + ":");
    for (FloatPoint p : a) b.append(fp(p)).append(";");
    return b.append("]").toString();
  }
  static String iiArr(int[][] a) {
    if (a == null) return "null";
    StringBuilder b = new StringBuilder("[" + a.length + ":");
    for (int[] p : a) b.append(p[0]).append("/").append(p[1]).append(";");
    return b.append("]").toString();
  }
  static String seg(LineSegment s) {
    if (s == null) return "null";
    return "SEG{" + ln(s.getStartClosingLine()) + " " + ln(s.getLine()) + " " + ln(s.getEndClosingLine()) + "}";
  }

  static Point[] genPoints(int lo, int hi, int c) {
    int n = r(lo, hi);
    Point[] pts = new Point[n];
    for (int i = 0; i < n; i++) pts[i] = new IntPoint(r(-c, c), r(-c, c));
    return pts;
  }

  static Line genLine(int c) {
    IntPoint a = new IntPoint(r(-c, c), r(-c, c));
    IntPoint b = new IntPoint(r(-c, c), r(-c, c));
    if (a.equals(b)) b = new IntPoint(b.x + 1, b.y);
    return new Line(a, b);
  }

  static IntBox genBox() {
    int x = r(-6, 6), y = r(-6, 6), w = r(1, 12), h = r(1, 12);
    return new IntBox(new IntPoint(x, y), new IntPoint(x + w, y + h));
  }

  public static void main(String[] args) {
    int iters = Integer.parseInt(args[0]);
    seed = Long.parseLong(args[1]);
    int mode = Integer.parseInt(args[2]);
    for (int it = 0; it < iters; it++) {
      if (mode == 3) {
        // targeted Polyline(Line[]) sweep: draw from a small pool of lines and their opposites
        int poolSize = r(2, 4);
        Line[] pool = new Line[poolSize * 2];
        for (int i = 0; i < poolSize; i++) { pool[2 * i] = genLine(4); pool[2 * i + 1] = pool[2 * i].opposite(); }
        int n = r(3, 10);
        final Line[] arr = new Line[n];
        for (int i = 0; i < n; i++) arr[i] = pool[r(0, pool.length - 1)];
        final Line[] arrCopy = arr.clone();
        emit("in", () -> { StringBuilder b = new StringBuilder("[" + arrCopy.length + ":"); for (Line l : arrCopy) b.append(ln(l)).append(";"); return b.append("]").toString(); });
        emit("fromLines", () -> pl(new Polyline(arr)));
        continue;
      }
      final int c = mode == 1 ? 9 : 6;
      final Point[] pts;
      if (mode == 4) {
        // 45/90-degree walk: the real trace geometry
        int n = r(3, 10);
        Point[] w = new Point[n];
        int x = r(-8, 8), y = r(-8, 8);
        int[] dx = {1, 1, 0, -1, -1, -1, 0, 1};
        int[] dy = {0, 1, 1, 1, 0, -1, -1, -1};
        for (int i = 0; i < n; i++) {
          int d = r(0, 7), len = r(1, 8);
          x += dx[d] * len; y += dy[d] * len;
          w[i] = new IntPoint(x, y);
        }
        pts = w;
      } else {
        pts = mode == 1 ? genPoints(1, 6, c) : genPoints(4, 12, c);
      }
      final int hw = mode == 1 ? r(0, 4) : r(1, 10);
      emit("pts", () -> ptArr(pts));
      final Polygon poly = new Polygon(pts);
      emit("polyCorners", () -> ptArr(poly.cornerArray()));
      emit("polyRevert", () -> ptArr(poly.revertCorners().cornerArray()));
      emit("polyWind", () -> String.valueOf(poly.windingNumberAfterClosing()));
      final Polyline p = new Polyline(pts);
      emit("p", () -> pl(p));
      emit("cornerCount", () -> String.valueOf(p.cornerCount()));
      emit("corners", () -> ptArr(p.corners()));
      emit("cornerApproxArr", () -> fpArr(p.cornerApproxArr()));
      emit("isEmpty", () -> String.valueOf(p.isEmpty()));
      emit("isPoint", () -> String.valueOf(p.isPoint()));
      emit("isOrth", () -> String.valueOf(p.isOrthogonal()));
      emit("is45", () -> String.valueOf(p.isMultipleOf45Degree()));
      emit("first", () -> cpt(p.firstCorner()));
      emit("last", () -> cpt(p.lastCorner()));
      emit("lenApprox", () -> f(p.lengthApprox()));
      final int a1 = r(0, 6), a2 = r(0, 6);
      emit("lenApproxAB", () -> f(p.lengthApprox(Math.min(a1, a2), Math.max(a1, a2))));
      emit("bbox", () -> box(p.boundingBox()));
      emit("bboxAB", () -> box(p.boundingBox(Math.min(a1, a2), Math.max(a1, a2))));
      emit("boctAB", () -> oct(p.boundingOctagon(Math.min(a1, a2), Math.max(a1, a2))));
      emit("reverse", () -> pl(p.reverse()));
      emit("offsetShapes", () -> tsArr(p.offsetShapes(hw)));
      emit("offsetShapesAB", () -> tsArr(p.offsetShapes(hw, Math.min(a1, a2), Math.max(a1, a2))));
      final int k = r(0, 5);
      emit("offsetShape", () -> ts(p.offsetShape(hw, k)));
      emit("offsetBox", () -> box(p.offsetBox(hw, k)));
      emit("translate", () -> pl(p.translateBy(new IntVector(r(-3, 3), r(-3, 3)))));
      final int fac = r(0, 7);
      emit("turn90", () -> pl(p.turn90Degree(fac, new IntPoint(r(-3, 3), r(-3, 3)))));
      emit("mirrorV", () -> pl(p.mirrorVertical(new IntPoint(r(-3, 3), r(-3, 3)))));
      emit("mirrorH", () -> pl(p.mirrorHorizontal(new IntPoint(r(-3, 3), r(-3, 3)))));
      final double ang = r(0, 12) * 0.5;
      emit("rotateApprox", () -> pl(p.rotateApprox(ang, new FloatPoint(r(-4, 4), r(-4, 4)))));
      final FloatPoint q = new FloatPoint(r(-12, 12), r(-12, 12));
      emit("nearest", () -> fp(p.nearestPointApprox(q)));
      emit("distance", () -> f(p.distance(q)));
      final IntPoint qi = new IntPoint(r(-12, 12), r(-12, 12));
      emit("contains", () -> String.valueOf(p.contains(qi)));
      emit("projection", () -> seg(p.projectionLine(qi)));
      emit("skipLines", () -> pl(p.skipLines(Math.min(a1, a2), Math.max(a1, a2))));
      final int nlc = r(0, 8);
      final double sl = r(1, 15);
      emit("shorten", () -> pl(p.shorten(nlc, sl)));
      final Point[] pts2 = mode == 1 ? genPoints(1, 6, c) : genPoints(4, 12, c);
      final Polyline p2 = new Polyline(pts2);
      emit("combine", () -> pl(p.combine(p2)));
      // combine with a polyline sharing an end corner
      final IntPoint sh1 = new IntPoint(r(-c, c), r(-c, c));
      final IntPoint sh2 = new IntPoint(r(-c, c), r(-c, c));
      emit("combineShared", () -> {
        Point[] pc = p.corners();
        if (pc.length == 0) return "skip";
        Point endPt = pc[pc.length - 1];
        Point[] q2 = new Point[] {endPt, sh1, sh2};
        return pl(p.combine(new Polyline(q2)));
      });
      final IntPoint sh3 = new IntPoint(r(-c, c), r(-c, c));
      final IntPoint sh4 = new IntPoint(r(-c, c), r(-c, c));
      emit("combineSharedStart", () -> {
        Point[] pc = p.corners();
        if (pc.length == 0) return "skip";
        Point startPt = pc[0];
        Point[] q2 = new Point[] {startPt, sh3, sh4};
        return pl(p.combine(new Polyline(q2)));
      });
      final int si = r(0, 6);
      final Line el = genLine(c);
      emit("split", () -> plArr(p.split(si, el)));
      final int lsn = r(0, 6);
      emit("lineSegment", () -> seg(new LineSegment(p, lsn)));
      emit("toPolyline", () -> { LineSegment s = new LineSegment(p, lsn); return pl(s.toPolyline()); });
      final IntBox bx = genBox();
      emit("shapeRotate", () -> ts(((TileShape) bx).rotateApprox(ang, new FloatPoint(r(-4, 4), r(-4, 4)))));
      emit("entrance", () -> iiArr(((TileShape) bx).entrancePoints(p)));
      emit("cutout", () -> plArr(((TileShape) bx).cutout(p)));
      final Simplex sx = bx.toSimplex();
      emit("simplexEntrance", () -> iiArr(((TileShape) sx).entrancePoints(p)));
      emit("simplexCutout", () -> plArr(((TileShape) sx).cutout(p)));
    }
    System.out.print(sb);
  }
}

File: scripts/differential/rust/src/bin/t16r.rs
use std::panic::{self, AssertUnwindSafe};

use copper_geometry::float_point::FloatPoint;
use copper_geometry::int_box::IntBox;
use copper_geometry::int_octagon::IntOctagon;
use copper_geometry::int_point::IntPoint;
use copper_geometry::int_vector::IntVector;
use copper_geometry::line::Line;
use copper_geometry::line_segment::LineSegment;
use copper_geometry::point::Point;
use copper_geometry::polygon::Polygon;
use copper_geometry::polyline::{Polyline, PolylineError};
use copper_geometry::simplex::Simplex;
use copper_geometry::tile_shape::TileShape;
use copper_geometry::vector::Vector;

struct Rng(u64);
impl Rng {
    fn next(&mut self) -> u64 {
        self.0 ^= self.0 << 13;
        self.0 ^= self.0 >> 7;
        self.0 ^= self.0 << 17;
        self.0
    }
    fn r(&mut self, lo: i32, hi: i32) -> i32 {
        let v = (self.next() >> 1) % ((hi - lo + 1) as u64);
        lo + v as i32
    }
}

fn f(d: f64) -> String { format!("b{}", d.to_bits() as i64) }
fn cpt(p: &Point) -> String {
    match p {
        Point::Int(q) => format!("I{}/{}", q.x, q.y),
        Point::Rational(r) => format!("R{}/{}/{}", r.x, r.y, r.z),
    }
}
fn ocpt(p: &Option<Point>) -> String { match p { None => "null".into(), Some(x) => cpt(x) } }
fn fp(p: &FloatPoint) -> String { format!("({},{})", f(p.x), f(p.y)) }
fn ofp(p: &Option<FloatPoint>) -> String { match p { None => "null".into(), Some(x) => fp(x) } }
fn ln(l: &Line) -> String { format!("[I{}/{};I{}/{}]", l.a.x, l.a.y, l.b.x, l.b.y) }
fn boxs(b: &IntBox) -> String { format!("B({},{},{},{})", b.ll.x, b.ll.y, b.ur.x, b.ur.y) }
fn octs(o: &IntOctagon) -> String {
    format!("O({},{},{},{},{},{},{},{})", o.left_x, o.bottom_y, o.right_x, o.top_y,
        o.upper_left_diagonal_x, o.lower_right_diagonal_x, o.lower_left_diagonal_x, o.upper_right_diagonal_x)
}
fn ts(s: &TileShape) -> String {
    match s {
        TileShape::Box(b) => boxs(b),
        TileShape::Octagon(o) => octs(o),
        TileShape::Simplex(sx) => {
            let mut b = String::from("S{");
            for i in 0..sx.border_line_count() {
                if i > 0 { b.push(' '); }
                b.push_str(&ln(&sx.border_line(i).unwrap()));
            }
            b.push('}');
            b
        }
    }
}
fn ts_arr(a: &[TileShape]) -> String {
    let mut b = format!("[{}:", a.len());
    for s in a { b.push_str(&ts(s)); b.push(';'); }
    b.push(']'); b
}
fn pl(p: &Polyline) -> String {
    let mut b = format!("P[{}:", p.lines().len());
    for l in p.lines() { b.push_str(&ln(l)); b.push(';'); }
    b.push(']'); b
}
fn plr(p: Result<Polyline, PolylineError>) -> String {
    match p { Ok(v) => pl(&v), Err(_) => "EXC".into() }
}
fn pl_arr(a: &[Polyline]) -> String {
    let mut b = format!("[{}:", a.len());
    for p in a { b.push_str(&pl(p)); b.push(';'); }
    b.push(']'); b
}
fn pt_arr(a: &[Point]) -> String {
    let mut b = format!("[{}:", a.len());
    for p in a { b.push_str(&cpt(p)); b.push(';'); }
    b.push(']'); b
}
fn fp_arr(a: &[FloatPoint]) -> String {
    let mut b = format!("[{}:", a.len());
    for p in a { b.push_str(&fp(p)); b.push(';'); }
    b.push(']'); b
}
fn ii_arr(a: &[[usize; 2]]) -> String {
    let mut b = format!("[{}:", a.len());
    for p in a { b.push_str(&format!("{}/{};", p[0], p[1])); }
    b.push(']'); b
}
fn seg(s: &LineSegment) -> String {
    format!("SEG{{{} {} {}}}", ln(&s.get_start_closing_line()), ln(&s.get_line()), ln(&s.get_end_closing_line()))
}
fn oseg(s: &Option<LineSegment>) -> String { match s { None => "null".into(), Some(x) => seg(x) } }

fn emit(out: &mut String, name: &str, t: impl FnOnce() -> String) {
    let v = match panic::catch_unwind(AssertUnwindSafe(t)) {
        Ok(v) => v,
        Err(_) => "EXC".to_string(),
    };
    out.push_str(name); out.push('='); out.push_str(&v); out.push('\n');
}

fn gen_points(rng: &mut Rng, lo: i32, hi: i32, c: i32) -> Vec<Point> {
    let n = rng.r(lo, hi);
    (0..n).map(|_| Point::Int(IntPoint::new(rng.r(-c, c), rng.r(-c, c)))).collect()
}
fn gen_line(rng: &mut Rng, c: i32) -> Line {
    let a = IntPoint::new(rng.r(-c, c), rng.r(-c, c));
    let mut b = IntPoint::new(rng.r(-c, c), rng.r(-c, c));
    if a == b { b = IntPoint::new(b.x + 1, b.y); }
    Line::new(a, b)
}
fn gen_box(rng: &mut Rng) -> IntBox {
    let x = rng.r(-6, 6); let y = rng.r(-6, 6); let w = rng.r(1, 12); let h = rng.r(1, 12);
    IntBox::from_coords(x, y, x + w, y + h)
}

fn main() {
    panic::set_hook(Box::new(|_| {}));
    let args: Vec<String> = std::env::args().collect();
    let iters: i32 = args[1].parse().unwrap();
    let seed: i64 = args[2].parse().unwrap();
    let mode: i32 = args[3].parse().unwrap();
    let mut rng = Rng(seed as u64);
    let mut out = String::new();
    for _ in 0..iters {
        if mode == 3 {
            let pool_size = rng.r(2, 4);
            let mut pool: Vec<Line> = Vec::new();
            for _ in 0..pool_size {
                let l = gen_line(&mut rng, 4);
                pool.push(l);
                pool.push(l.opposite());
            }
            let n = rng.r(3, 10);
            let arr: Vec<Line> = (0..n).map(|_| pool[rng.r(0, pool.len() as i32 - 1) as usize]).collect();
            let a2 = arr.clone();
            emit(&mut out, "in", || {
                let mut b = format!("[{}:", a2.len());
                for l in &a2 { b.push_str(&ln(l)); b.push(';'); }
                b.push(']'); b
            });
            emit(&mut out, "fromLines", || plr(Polyline::from_lines(arr)));
            continue;
        }
        let c = if mode == 1 { 9 } else { 6 };
        let pts: Vec<Point> = if mode == 4 {
            let n = rng.r(3, 10);
            let mut x = rng.r(-8, 8); let mut y = rng.r(-8, 8);
            let dx = [1, 1, 0, -1, -1, -1, 0, 1];
            let dy = [0, 1, 1, 1, 0, -1, -1, -1];
            let mut w = Vec::new();
            for _ in 0..n {
                let d = rng.r(0, 7) as usize; let len = rng.r(1, 8);
                x += dx[d] * len; y += dy[d] * len;
                w.push(Point::Int(IntPoint::new(x, y)));
            }
            w
        } else if mode == 1 { gen_points(&mut rng, 1, 6, c) } else { gen_points(&mut rng, 4, 12, c) };
        let hw = if mode == 1 { rng.r(0, 4) } else { rng.r(1, 10) };
        emit(&mut out, "pts", || pt_arr(&pts));
        let poly = Polygon::new(pts.clone());
        emit(&mut out, "polyCorners", || pt_arr(poly.corner_array()));
        emit(&mut out, "polyRevert", || pt_arr(poly.revert_corners().corner_array()));
        emit(&mut out, "polyWind", || poly.winding_number_after_closing().to_string());
        let p = Polyline::from_points(&pts);
        emit(&mut out, "p", || pl(&p));
        emit(&mut out, "cornerCount", || p.corner_count().to_string());
        emit(&mut out, "corners", || pt_arr(&p.corners()));
        emit(&mut out, "cornerApproxArr", || fp_arr(&p.corner_approx_arr()));
        emit(&mut out, "isEmpty", || p.is_empty().to_string());
        emit(&mut out, "isPoint", || p.is_point().to_string());
        emit(&mut out, "isOrth", || p.is_orthogonal().to_string());
        emit(&mut out, "is45", || p.is_multiple_of_45_degree().to_string());
        emit(&mut out, "first", || ocpt(&p.first_corner()));
        emit(&mut out, "last", || ocpt(&p.last_corner()));
        emit(&mut out, "lenApprox", || f(p.length_approx()));
        let a1 = rng.r(0, 6); let a2v = rng.r(0, 6);
        let (alo, ahi) = (a1.min(a2v) as usize, a1.max(a2v) as usize);
        emit(&mut out, "lenApproxAB", || f(p.length_approx_between(alo, ahi)));
        emit(&mut out, "bbox", || boxs(&p.bounding_box()));
        emit(&mut out, "bboxAB", || boxs(&p.bounding_box_between(alo, ahi)));
        emit(&mut out, "boctAB", || octs(&p.bounding_octagon_between(alo, ahi)));
        emit(&mut out, "reverse", || plr(p.reverse()));
        emit(&mut out, "offsetShapes", || ts_arr(&p.offset_shapes(hw)));
        emit(&mut out, "offsetShapesAB", || ts_arr(&p.offset_shapes_between(hw, alo, ahi)));
        let k = rng.r(0, 5) as usize;
        emit(&mut out, "offsetShape", || match p.offset_shape(hw, k) { None => "null".into(), Some(s) => ts(&s) });
        emit(&mut out, "offsetBox", || match p.offset_box(hw, k) { None => "null".into(), Some(b) => boxs(&b) });
        let tv = IntVector::new(rng.r(-3, 3), rng.r(-3, 3));
        emit(&mut out, "translate", || plr(p.translate_by(&Vector::Int(tv))));
        let fac = rng.r(0, 7);
        let tp = IntPoint::new(rng.r(-3, 3), rng.r(-3, 3));
        emit(&mut out, "turn90", || plr(p.turn_90_degree(fac, &tp)));
        let mv = IntPoint::new(rng.r(-3, 3), rng.r(-3, 3));
        emit(&mut out, "mirrorV", || plr(p.mirror_vertical(&mv)));
        let mh = IntPoint::new(rng.r(-3, 3), rng.r(-3, 3));
        emit(&mut out, "mirrorH", || plr(p.mirror_horizontal(&mh)));
        let ang = rng.r(0, 12) as f64 * 0.5;
        let rp = FloatPoint::new(rng.r(-4, 4) as f64, rng.r(-4, 4) as f64);
        emit(&mut out, "rotateApprox", || pl(&p.rotate_approx(ang, &rp)));
        let q = FloatPoint::new(rng.r(-12, 12) as f64, rng.r(-12, 12) as f64);
        emit(&mut out, "nearest", || ofp(&p.nearest_point_approx(&q)));
        emit(&mut out, "distance", || f(p.distance(&q)));
        let qi = Point::Int(IntPoint::new(rng.r(-12, 12), rng.r(-12, 12)));
        emit(&mut out, "contains", || p.contains(&qi).to_string());
        emit(&mut out, "projection", || oseg(&p.projection_line(&qi)));
        emit(&mut out, "skipLines", || plr(p.skip_lines(alo, ahi)));
        let nlc = rng.r(0, 8) as usize;
        let sl = rng.r(1, 15) as f64;
        emit(&mut out, "shorten", || plr(p.shorten(nlc, sl)));
        let pts2 = if mode == 1 { gen_points(&mut rng, 1, 6, c) } else { gen_points(&mut rng, 4, 12, c) };
        let p2 = Polyline::from_points(&pts2);
        emit(&mut out, "combine", || match p.combine(&p2) { Ok(v) => pl(&v), Err(_) => "EXC".into() });
        let sh1 = IntPoint::new(rng.r(-c, c), rng.r(-c, c));
        let sh2 = IntPoint::new(rng.r(-c, c), rng.r(-c, c));
        emit(&mut out, "combineShared", || {
            let pc = p.corners();
            if pc.is_empty() { return "skip".into(); }
            let end_pt = pc[pc.len() - 1].clone();
            let q2 = vec![end_pt, Point::Int(sh1), Point::Int(sh2)];
            match p.combine(&Polyline::from_points(&q2)) { Ok(v) => pl(&v), Err(_) => "EXC".into() }
        });
        let sh3 = IntPoint::new(rng.r(-c, c), rng.r(-c, c));
        let sh4 = IntPoint::new(rng.r(-c, c), rng.r(-c, c));
        emit(&mut out, "combineSharedStart", || {
            let pc = p.corners();
            if pc.is_empty() { return "skip".into(); }
            let start_pt = pc[0].clone();
            let q2 = vec![start_pt, Point::Int(sh3), Point::Int(sh4)];
            match p.combine(&Polyline::from_points(&q2)) { Ok(v) => pl(&v), Err(_) => "EXC".into() }
        });
        let si = rng.r(0, 6) as usize;
        let el = gen_line(&mut rng, c);
        emit(&mut out, "split", || match p.split(si, &el) { Ok(None) => "null".into(), Ok(Some(a)) => pl_arr(&a), Err(_) => "EXC".into() });
        let lsn = rng.r(0, 6) as usize;
        emit(&mut out, "lineSegment", || oseg(&LineSegment::from_polyline(&p, lsn)));
        emit(&mut out, "toPolyline", || match LineSegment::from_polyline(&p, lsn) { None => "EXC".into(), Some(s) => match s.to_polyline() { Ok(pl_) => pl(&pl_), Err(_) => "EXC".into() } });
        let bx = gen_box(&mut rng);
        let srp = FloatPoint::new(rng.r(-4, 4) as f64, rng.r(-4, 4) as f64);
        emit(&mut out, "shapeRotate", || ts(&TileShape::Box(bx).rotate_approx(ang, &srp)));
        emit(&mut out, "entrance", || ii_arr(&TileShape::Box(bx).entrance_points(&p)));
        emit(&mut out, "cutout", || match TileShape::Box(bx).cutout_polyline(&p) { Ok(v) => pl_arr(&v), Err(_) => "EXC".into() });
        let sx: Simplex = bx.to_simplex();
        emit(&mut out, "simplexEntrance", || ii_arr(&TileShape::Simplex(sx.clone()).entrance_points(&p)));
        emit(&mut out, "simplexCutout", || match TileShape::Simplex(sx.clone()).cutout_polyline(&p) { Ok(v) => pl_arr(&v), Err(_) => "EXC".into() });
    }
    print!("{}", out);
}
